dimod/binary_quadratic_model.py

```diff
@@ -1612,6 +1612,92 @@ def from_json(cls, obj):
 
         return json.load(obj,  object_hook=bqm_decode_hook)
 
+    def to_bson(self, fp=None):
+        """Convert a binary quadratic model to an effecient BSON format.
+
+        Args:
+            fp (file, optional):
+                A `.write()`-supporting `file object`_ to save the binary 
+                quadratic model encoded in BSON format to. If not provided, 
+                returns bytes.
+        """
```

Comment:
Return value type should **never** depend on argument value. Besides being blasphemous, it's also asking for trouble later on, making API too "magical".

Instead, you should require `fp`, and if a user wants a byte stream instead of file, she can use [`io.BytesIO`](https://docs.python.org/3/library/io.html#io.BytesIO):

```
import io
buf = io.BytesIO()
bqm.to_bson(buf)
```

Reply:
I wholeheartedly agree with your first statement. I chose to follow the interface of `to_json`/`from_json` as closely as possibly for consistency. 

I would prefer to keep `fp` optional because though passing in a buffer works fine, I think it's a little more clunky to use than these methods should be. 

Instead I propose always returning `bytes`, and if `fp` is provided, also writing those bytes to the file. If the user just cares about the file, they can ignore the return value. Thoughts?

Reply:
I don't think I agree so strongly, as is in evidence from `to_json/from_json`, but I am happy enough to make the change for the 0.7.0 release.

In terms of return value, I think my instinct is to always return bytes and the user can then write them to a file if they wish
```
bytes_ = bqm.to_bson()
with open('file') as fp:
    fp.write(bytes_)
```

Reply:
We can then make `to_json` return a json-serializable string always so the pattern would be
```
with open('file') as fp:
    fp.write(json.dumps(bqm.to_json())
```
Which might be less efficient but would be much simpler

Reply:
Unless the argument for passing in `fp` is that the memory footprint is less?

Reply:
Good point about local consistency. It's more important than global. (At least according to PEP-8.)

But I don't see your point on clunkiness. One might say it's clunky to need to open a file before calling `to_bson(fp)`. I'm not saying we should support filenames/paths as arguments, but that *there's beauty in having a method that does one thing, and does it well*.

Honestly, I don't see any benefit of supporting `fp`, i.e.:
```
with open('path') as fp:
    fp.write(bqm.to_bson())
```
vs.
```
with open('path') as fp:
    bqm.to_bson(fp)
```
On contrary, by not accepting `fp`, we have better separation of concerns.